Docs for palettes and aes (#108)
posit-dev/ggsql: doc/syntax/scale/aesthetic/color.qmd

---

Color values are used in two places in ggsql: When setting stroke color and fill color. Stroke color governs the color of lines, be it the outline of a rectangle (e.g. in a [bar layer](../../layer/bar.qmd)), or the line of a path (e.g. in a [line layer](../../layer/line.qmd)). The fill color determines the color of the inside of a shape. It follows that all layers respond to the stroke aesthetic but only those containing closed shapes respond to the fill aesthetic.

## The color meta-aesthetic
In addition to the `fill` and `stroke` aesthetics there is also a `color` meta-aesthetic. Setting this will set the fill *and* stroke at the same time unless they are set directly. For instance, the following will set the fill and stroke to red

```ggsql
DRAW bar
    SETTING color => 'red'
```

whereas this will set the fill to red and stroke to black

```ggsql
DRAW bar
    SETTING color => 'red', stroke => 'black'
```

The same logic applies when defining a scale. When you define a `color` scale two scales will be created, one for `fill` and one for `stroke` (unless either are defined explicitly) with the same settings. For instance, the following

```ggsql
SCALE color TO viridis
```

> Will define two separate scales for `fill` and `stroke` that both uses the `viridis` palette.

Reviewer: ```suggestion
defines two separate scales for `fill` and `stroke` that both use the `viridis` palette.
```

In general it is highly advisable to be explicit and use the `fill` and `stroke` aesthetics directly, but the `color` meta-aesthetic is there as a quick shortcut if you need it.

## Literal values

> Literal color values can be defined in any CSS compatible way. For a complete overview see the [documentation for the csscolorparser crate](https://docs.rs/csscolorparser/latest/csscolorparser/) which is what we use internally. The gist of it is that colors can be defined in one of three different ways:

Reviewer: I think this is one of those details we can hide away in a footnote
```suggestion
Literal color values can be defined in any CSS compatible way ^[For a complete overview see the [documentation for the csscolorparser crate](https://docs.rs/csscolorparser/latest/csscolorparser/) which is what we use internally.]. The gist of it is that colors can be defined in one of three different ways:
```

Author: Good call

* As a named color, e.g. `'red'`, or `'lemonchiffon'`
* As a hex color value, e.g. `'#FF0000'` or `'#FFFACD'`

> * As a css color function, e.g. `'rgb(100%, 0%, 0%)'` or `'oklab(98%, -0.012, 0.057)'`

Reviewer: Might be best to foreshadow comment on the `opacity`
```suggestion
* As a hex color value, e.g. `'#FF0000'` or `'#FFFACD88'`
* As a css color function, e.g. `'rgba(100%, 0%, 0%, 50%)'` or `'oklab(98%, -0.012, 0.057)'`
```

> Literal color values are used in two places in ggsql. Either when setting an aesthetic (as opposed to mapping), e.g.

Reviewer: I think there is a 3rd way with identity scales

```ggsql
DRAW line
    SETTING stroke => '#FF0000'
```

Or when you define the output range of a fill or stroke scale manually, e.g.

```ggsql
SCALE fill TO ['rgb(100%, 0%, 0%)', 'lemonchiffon']
```

Note that all different types of color notation can be mixed in the same place.

Literal color values may be translucent, either by providing a fourth channel the the hex-notation, or by using a css color function that includes an alpha level (e.g. `rgba()`). You should avoid mixing this with the use of the opacity aesthetic to ensure the opacity is predictable.

> ## Palettes

Reviewer: I think the palettes look great with the gradients/blocks. This section is a bit tricky to navigate. Let's suppose you quickly want to see the built-in discrete palettes; you can't easily click your way to there, you have to scroll through a bunch of stuff you're not looking for. I'm not sure if we could increase the TOC depth or include some clickable intradocument links or something.

Author: I've increased the depth of the toc - I agree that for color the navigation suffers from all the palettes

|------|----------|--------|-------------|
| `ggsql10` | ![](examples/swatch_ggsql10.svg){height=20 .lightbox} | 10 | **Default.** Optimized for distinguishability |

#### Tableau

Tableau's categorical palette, widely used in data visualization tools.

::: {.callout-note}
**Source:** Tableau Software. [tableau.com](https://www.tableau.com)
:::

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `tableau10` | ![](examples/swatch_tableau10.svg){height=20 .lightbox} | 10 | Tableau's default categorical palette |

#### D3

D3's Category 10 palette, commonly used in web-based visualizations.

::: {.callout-note}
**Source:** Bostock, M. D3.js. [d3js.org](https://d3js.org)
:::

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `category10` | ![](examples/swatch_category10.svg){height=20 .lightbox} | 10 | D3's default categorical palette |

#### ColorBrewer Qualitative

Qualitative palettes from ColorBrewer, designed for categorical data where no ordering is implied.

::: {.callout-note}
**Source:** Brewer, C.A. (2002). ColorBrewer: Color Advice for Cartography. [colorbrewer2.org](https://colorbrewer2.org)
:::

> ##### Set Palettes

Reviewer: Even if the comment about not sorting by source is dismissed, I think all of these colorbrewer discrete palettes should be in one single table instead of separate tables for 1-3 palettes.

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `dark2` | ![](examples/swatch_dark2.svg){height=20 .lightbox} | 8 | Dark, saturated colors |

##### Paired Palette

Colors arranged in light-dark pairs, useful for showing related categories.

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `paired` | ![](examples/swatch_paired.svg){height=20 .lightbox} | 12 | Light-dark paired colors |

##### Accent Palette

A palette designed for highlighting specific categories.

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `accent` | ![](examples/swatch_accent.svg){height=20 .lightbox} | 8 | Accent colors for emphasis |

#### Kelly's Colors of Maximum Contrast

A scientifically derived palette designed for maximum visual distinction between colors.

::: {.callout-note}
**Source:** Kelly, K.L. (1965). Twenty-two colors of maximum contrast. *Color Engineering*, 3(6), 26-27.
:::

| Name | Swatches | Colors | Description |
|------|----------|--------|-------------|
| `kelly22` | ![](examples/swatch_kelly22.svg){height=20 .lightbox} | 20 | Maximum contrast colors |

> ### Choosing a Palette

Reviewer: The diagram is very much not a priority right now
